Fixed #26291 -- Allowed loaddata to handle forward references in natural_key fixtures.

inglesp opened the pull request:
This is a first stab at supporting forward references in fixtures that
use natural foreign keys.  See [ticket 26291](https://code.djangoproject.com/ticket/26291)

Before this commit, the following fixture could not be deserialized by
loaddata if an author whose name is "John Steinbeck" did not already
exist:

```
[
  {
    "model": "app.book",
    "fields": {
      "title": "East Of Eden",
      "author": ["John Steinbeck"]
    }
  },
  {
    "model": "app.author",
    "fields": {
      "name": "John Steinbeck",
      "date_of_birth": "1902-02-27"
    }
  }
]
```

This problem can be avoided by careful ordering of fixtures, but it is
unavoidable for fixtures that contain circular references.

With this commit, when loaddata tries to deserialize a fixture
containing an object that has a natural foreign key to another object
that does not yet exist, it will defer setting the foreign key attribute
until after all fixtures are loaded.

TODO: documentation

freakboy3742 commented on django/core/serializers/python.py:
```diff
@@ -91,6 +92,7 @@ def Deserializer(object_list, **options):
     """
     db = options.pop('using', DEFAULT_DB_ALIAS)
     ignore = options.pop('ignorenonexistent', False)
+    handle_forward_references = options.pop('handle_forward_references', False)
```
What's the use case for *not* handling forward references? In particular, what's the use case for making this user configurable?

inglesp replied:
I put this in for backwards compatibility, but it's probably not necessary -- I don't see why anybody would be relying on the current behaviour of being unable to load forward references.

inglesp replied:
I've thought about this a little further.  I think this is required, since automatically handling forward references changes the behaviour of user code.

In particular, somebody might be relying on an exception being raised when they try to deserialize data that contains forward references.  The new behaviour won't raise an exception, and instead the same user code will create incomplete objects in the database.

This is because deserialization now requires two steps: firstly, we try to deserialize some data, and secondly, we have to call `save_deferred_fields` on any deserialized objects that had forward references.

Is this an argument to add some new code, to be used by `loaddata` and by user code, that removes this foot-gun?

inglesp replied:
That said, some new code wouldn't remove the foot-gun, it'd just make it easier to avoid the foot-gun.

timgraham replied:
To prevent incomplete objects, is there a way to detect the missing forward references and then raise an exception that would cause loaddata to rollback the transaction?

inglesp replied:
Hi @timgraham -- I'm afraid I don't understand your point.  loaddata is in control of how it deserializes objects, and so with my changes it shouldn't ever need to roll back a transaction to deal with forward references.

I've added some documentation in dced428e1248bb73cc1262ea16952fbc45e2aa27.  What's your opinion on the current state of this work?

inglesp replied:
I still think that https://github.com/django/django/pull/6221#r68929405 is correct.  That is, automatically handling forward references changes behaviour and so `handle_forward_references` is required to maintain backwards compatibility.

The problem is that `serializers.deserialize` does not save data to the database, and so there's no point that `serializers.deserialize` can check to see whether there are incomplete objects -- this must be done by the caller of `serializers.deserialize`.

freakboy3742 commented on django/core/serializers/python.py:
```diff
@@ -142,12 +145,17 @@ def m2m_convert(v):
-                        m2m_data[field.name].append(m2m_convert(pk))
+                        m2m_field_values.append(m2m_convert(pk))
                 except Exception as e:
+                    if isinstance(e, ObjectDoesNotExist) and handle_forward_references:
```
This seems a bit convoluted - is there a reason this isn't just a ObjectDoesNotExist exception handler?

inglesp replied:
Agree it's not very nice -- but we only want to handle `ObjectDoesNotExist` when `handle_forward_references` is `True`.  This goes away if we decide we do always want to handle forward references.

freakboy3742 commented on django/core/serializers/base.py:
```diff
@@ -207,6 +208,49 @@ def save(self, save_m2m=True, using=None, **kwargs):
+
+        self.m2m_data = {}
+
+        for field, field_value in self.deferred_fields.items():
```
Isn't this block largely duplicated code? Shouldn't it be possible to factor out the "set the value of this m2m/m2o field" logic into a common location so that there's no duplication between first and second pass updates?

inglesp replied:
It's only quite similar.  I'll look into whether factoring this out makes the code easier to work with.

timgraham commented:
@inglesp -- will you return to this soon or should we close this for now?

inglesp commented:
Hi @timgraham, thanks for the nudge.

I've responded to the one outstanding question [here](https://github.com/django/django/pull/6221#discussion_r68929405).

Should I continue discussion here, or on the ticket?

timgraham commented:
You could try writing to the [django-developers mailing list](https://groups.google.com/forum/#!forum/django-developers), that's usually the best place for discussion on design decisions.

milind7705 commented:
When will this be merged in the master? I am facing similar issue of forward references where my foreign key in the model points to self and couldn;t find the help on how to resolve the same without the django code change. 
Pointers will be appreciated.

class Person(models.Model):
    person_name = models.CharField(max_length=128, null=False, blank=False, default="")
    description = models.CharField(max_length=512, null=False, blank=False, default="")
    failed person_name = models.ForeignKey("self", on_delete=models.CASCADE, null=True, blank=True)
    failed person_name.verbose_name = "Failure Person"

    def __unicode__(self):
        return self.person_name

    def __eq__(self, other):
        if isinstance(other, Person):
            return self.person_name == other.person_name

    def __ne__(self, other):
        if isinstance(other, Person):
            return not self.person_name == other.person_name

inglesp commented:
Hi @carltongibson!  Thanks for picking this up.  I've ticked the box so that you can push your changes, and I'll aim to respond to your questions in the next week.

carltongibson commented:
@inglesp OK. Updated. (Please take a look when you have a moment to make sure it still looks how you expect.) 

Let's see if we can get this over the line 🙂. Thanks for the effort!

inglesp commented:
> If I read the history right, you settled on needing the handle_forward_references flag in order to maintain backward compatibility. But we're not exposing that to the loaddata command. Do we need to? (If not do we need to add a class attribute or something so that the command could be subclassed? Do we need docs for that? Or are we expecting people who need the old behaviour to implement their own command? If none of the above, why are we keeping the option?)

We don't need to expose the `handle_forward_references` argument to `loaddata`, since this PR changes the `loaddata` code to pass `handle_forward_references=True` to `serializers.deserialize()`.

Users only need to know/care about `handle_forward_references` when calling `serializers.deserialize()` directly.

> There's Russel's point about refactoring save_deferred_fields in serializers/base.py. Did you want to address this?

I've tweaked it slightly in 8ebf8c5397066fec55d75684962d5f039af87acc.  Could it be further improved do you think?

> Doc changes look like a good start. (Needed: release notes, Depending on 1, loaddata docs, etc)

I've added to the release notes in 8ebf8c5397066fec55d75684962d5f039af87acc.  Is anything else required?

carltongibson commented:
> Users only need to know/care about handle_forward_references when calling serializers.deserialize() directly.

Yes. Fine. (The BC use-case was users catching the existing exception to handle this kind of case. They need to do that from a script, rather than `loaddata`. The new behaviour is clearly better in other cases.) 

Thanks for the effort @inglesp. This looks great to me.

timgraham commented on docs/topics/serialization.txt:
```diff
@@ -514,17 +514,67 @@ command line flags to generate natural keys.
+
+Natural keys and forward references
+-----------------------------------
+
```
Add `.. versionadded:: 2.1`

timgraham commented on django/core/serializers/base.py:
```diff
@@ -218,6 +219,46 @@ def save(self, save_m2m=True, using=None, **kwargs):
 
+    def save_deferred_fields(self, using=None):
+        if self.deferred_fields is None:
+            return
```
According to the coverage report, this line isn't executed.

timgraham commented on django/core/serializers/base.py:
```diff
@@ -218,6 +219,46 @@ def save(self, save_m2m=True, using=None, **kwargs):
+                    raise DeserializationError.WithData(e, label, self.object.pk, field_value)
+                value = getattr(obj, field_name)
+                if model._meta.pk.remote_field:
+                    value = value.pk
```
According to the coverage report, this line isn't executed.

timgraham commented on django/core/serializers/base.py:
```diff
@@ -218,6 +219,46 @@ def save(self, save_m2m=True, using=None, **kwargs):
+                setattr(self.object, field.attname, value)
+
+            else:
+                assert False
```
I would either remove this or add a helpful exception message.

timgraham commented on django/core/management/commands/loaddata.py:
```diff
@@ -163,7 +168,7 @@ def load_label(self, fixture_label):
                 objects = serializers.deserialize(
                     ser_fmt, fixture, using=self.using, ignorenonexistent=self.ignore,
-                )
+                    handle_forward_references=True)
```
There  should be a trailing comma and the parenthesis goes on the next line as it was before.

inglesp commented:
Thanks @timgraham for the review.

I've addressed all your comments -- in most cases they were trivial, but it turned out the best way to address the note in https://github.com/django/django/pull/6221#discussion_r174977139 about missing coverage was to do some refactoring so that code is shared between `django.core.serializers.python.Deserializer` and `django.core.serializers.base.DeserializedObject.save_deferred_fields`.

(This is exactly what Russ had suggested two years ago in https://github.com/django/django/pull/6221#discussion-diff-55210932...)

I'm sorry that this creates more review work!

timgraham commented:
Thanks. Wwould it be possible to do some of the the refactoring (moving existing code) as a separate PR that precedes this one so it's easier to see the changes that are necessary for the rest of this?

inglesp commented:
Good idea.  Have run out of time this evening, but should have time tomorrow.

inglesp commented:
I've created #9799 and #9800.  Let me know if it's not clear.

timgraham commented:
Did you answer my question about if this applicable to the XML serializer?

inglesp commented:
> Did you answer my question about if this applicable to the XML serializer?

Er, no, sorry.

> Is this applicable to the XML serializer?

Yes, but as far as I can tell, no code changes are required.

timgraham commented:
Could some tests be added to be sure? It might be that tests in `tests/serializers/test_natural.py` or similar would ease creating tests that run across all serializers. I guess at least one of the existing tests for `loaddata` should remain for the change to that command.

inglesp commented:
Apologies for sitting on this for so long.  It's not abandoned!  I just have limited time at the moment, and *surprise*, adding a test to `serializers.test_natural.NaturalKeySerializerTests` finds that this code doesn't work for XML (but it does for all other formats) so I've got some more digging to do.

inglesp commented:
This PR now handles XML serialization as well, and this is now tested.

auvipy commented on docs/releases/2.1.txt:
```diff
@@ -204,6 +204,19 @@ Requests and Responses
   wants to download the file. ``FileResponse`` also tries to set the
   ``Content-Type`` and ``Content-Length`` headers where appropriate.
 
```
this section should be moved to 2.2.txt

auvipy commented on docs/topics/serialization.txt:
```diff
@@ -514,17 +514,69 @@ command line flags to generate natural keys.
+Natural keys and forward references
+-----------------------------------
+
+.. versionadded:: 2.1
```
2.2

inglesp commented:
Woohoo, thanks @timgraham and @carltongibson!

(And thanks @auvipy for taking a look.)
